packages/lib/gpcircuits/src/artifacts.ts

```diff
@@ -0,0 +1,91 @@
+import * as fastfile from "fastfile";
+import { CircuitArtifactPaths, CircuitDesc } from "./types";
+
+/**
+ * Represents a pre-loaded verification key to be passed to SnarkJS, as loaded
+ * by {@link loadVerificationKey}.
+ */
+export type VerificationKey = object;
+
+/**
+ * Loads a verification key from a JSON file.
+ *
+ * @param vkeyPath path to load artifacts, which could be a URL (in browser)
+ *   or a file path (in server or utests).
+ * @returns the verification key as an object
+ * @throws if file access or JSON parsing fails
+ */
+export async function loadVerificationKey(
+  vkeyPath: string
+): Promise<VerificationKey> {
+  // This method of binary loading files using fastfile is the same as what is
+  // used in snarkjs to load pkey and wasm artifacts.  It works for local file
+  // paths (in Node) as well as URLs (in browser).
+  // The string decoding and JSON parsing is specific to our use of vkeys, which
+  // snarkjs assumes are already in memory via import.
+  let fd: fastfile.FastFile | undefined = undefined;
+  try {
+    fd = await fastfile.readExisting(vkeyPath);
+    const bytes = await fd.read(fd.totalSize);
+    return JSON.parse(Buffer.from(bytes).toString("utf8"));
+  } finally {
+    if (fd !== undefined) {
+      await fd.close();
+    }
+  }
+}
+
+/**
+ * Determines the right path for loading circuit artifacts for a given
+ * circuit.
+ *
+ * @param root root path to load artifacts, which could be a URL (in browser)
+ *   or a file path (in server or utests).
+ * @param cd description of the GPC circuit
+ * @returns collection of artifact paths
+ */
+export function gpcArtifactPaths(
+  root: string,
+  cd: CircuitDesc
+): CircuitArtifactPaths {
+  if (!root.endsWith("/")) {
+    root = root + "/";
+  }
```

Comment:
generally i use `url-join` package for stuff like this to never have to think about the `'/'` character

Reply:
This might be a URL or it might be a file path.  I originally used path.join and discovered it did bad things when given a URL.  Does url-join do bad things with file paths?  Seems likely if it handles relative URLs properly.